tests/unit/obuf-limits.tcl

```diff
@@ -18,117 +18,65 @@ start_server {tags {"obuf-limits"}} {
+            set omem 0
+            set start_time 0
+            set time_elapsed 0
+            while 1 {
+                r publish foo [string repeat "x" 1000]
+                set clients [split [r client list] "\r\n"]
+                set c [lsearch -inline $clients *name=test_client*]
+                if {$wait_for_timeout && $c == ""} {
+                    # Make sure we're disconnected when we reach the soft limit
+                    assert {$omem >= 100000 && $time_elapsed >= $soft_limit_time}
+                    break
+                } else {
+                    assert {[regexp {omem=([0-9]+)} $c - omem]}
+                }
+                if {$omem > 100000} {
+                    if {$start_time == 0} {set start_time [clock milliseconds]}
+                    set time_elapsed [expr {[clock milliseconds]-$start_time}]
+                    # Make sure test isn't taking too long
+                    assert {$time_elapsed <= [expr $soft_limit_time+2000]}
+                    if {!$wait_for_timeout && $time_elapsed >= [expr $soft_limit_time-1000]} break
+                    # Slow down loop when omem has reached the limit.
+                    after 10
+                } else {
+                    # if the OS socket buffers swallowed what we previously filled, reset the start timer.
+                    set start_time 0
+                }
             }
-        }
-        assert {$omem >= 100000 && $time_elapsed < 6}
-        $rd1 close
-    }
-
-    test {Client output buffer soft limit is enforced when no traffic} {
-        r config set client-output-buffer-limit {pubsub 0 100000 6}
-        set rd1 [redis_deferring_client]
-
-        $rd1 client setname test_client
-        set reply [$rd1 read]
-        assert {$reply eq "OK"}
-
-        $rd1 subscribe foo
-        set reply [$rd1 read]
-        assert {$reply eq "subscribe foo 1"}
 
-        set omem 0
-        set start_time 0
-        set time_elapsed 0
-        while 1 {
-            r publish foo [string repeat "x" 1000]
-            set clients [split [r client list] "\r\n"]
-            set c [lsearch -inline $clients *name=test_client*]
-            assert {[regexp {omem=([0-9]+)} $c - omem]}
-            if {$omem > 100000} {
-                if {$start_time == 0} {set start_time [clock seconds]}
-                set time_elapsed [expr {[clock seconds]-$start_time}]
-                if {$time_elapsed >= 3} break
-                # Slow down loop when omem has reached the limit.
-                after 10
-            } else {
-                # if the OS socket buffers swallowed what we previously filled, reset the start timer.
-                set start_time 0
+            if {!$wait_for_timeout} {
+                # Wait for soft limit to time out
+                after [expr {$soft_limit_time+1500 - ([clock milliseconds]-$start_time)}]
+                set clients [split [r client list] "\r\n"]
+                assert {[lsearch $clients *name=test_client*] == -1}
```

Comment:
let's convert that to a wait_for_condition loop, in many cases it'll finish sooner, but in case it doesn't we'll have a longer timeout

Reply:
Done, shaved ~1sec from the test.